server/chat-plugins/net-filters.ts

```diff
@@ -309,14 +346,40 @@ export const commands: ChatCommands = {
 			}
 			const targetPath = FS(`logs/chat/${roomid}/${date.slice(0, -3)}/${date}.txt`);
 			if (!targetPath.existsSync()) return this.errorReply(`Logs for that date not found`);
-			this.sendReply(`Initating training...`);
-			const response = await PM.query({data: {path: targetPath.path, result}, type: 'trainfrom'});
+			this.privateGlobalModAction(`${user.name} used /netfilter trainfrom ${roomid} (${date})`);
+			const response = await NeuralNetChecker.query({data: {path: targetPath.path, result}, type: 'trainfrom'});
 			this.sendReply(`Training completed in ${Chat.toDurationString(response[0])}`);
 			this.privateGlobalModAction(
 				`${user.name} trained the net filters on logs from ${roomid} (${date} - ${Chat.count(response[1], 'lines')})`
 			);
 			this.stafflog(`Result: ${result} | Time: ${response[0]}ms`);
 		},
+		cancel: 'stop',
+		stop(target, room, user) {
+			checkAllowed(this);
+			let count = 0;
+			const PM = processes.training;
+			const running = PM.processes.filter(p => p.load > 0);
+			if (!running.length) {
+				return this.errorReply(`No train tasks are pending`);
+			}
+			for (const subProcess of running) {
+				for (const [task, resolve] of subProcess.pendingTasks) {
+					resolve({error: `The pending train query was cancelled.`});
+					subProcess.pendingTasks.delete(task);
+					count++;
+				}
+				// kill is done here to ensure the process is actually killed right now
+				// fsr the disconnect() done in destroy doesn't stop the running task
+				subProcess.destroy();
+				subProcess.process.removeAllListeners('disconnect'); // if we don't do this it thinks it's a crash
+				subProcess.process.kill();
+				PM.processes.splice(PM.processes.indexOf(subProcess), 1);
+			}
+			PM.spawn(NUM_PROCESSES['training']);
```

Comment:
These should just use `PMTraining` directly. Also why are you manually destroying the processes?

Reply:
It feels like a lot of this code goes inside `lib/process-manager.ts`